### ISSUE TITLE
ARP issue 

### ISSUE THREAD

**Issue by Mutantt** [projectcalico/calico#4915]
Hi
We have a weird issue, when I create a pod it can't ping the IP of worker it belongs to but it can ping other workers also other pods rather than they located in the same worker or not. I changed  natoutgoing from true to false and the issue was the same. I checked proxy_arp and it's been set to 1. also we have default route on our hosts. I can ping my pod from the worker that it belongs to successfully. when I ping in both directions(ping node from inside the pod and vice versa at the same time) ping is successful from pod. but if  I cancel the ping from worker, ping from pod also fail.
would you please advice what can I do to solve the problem?

**Comment by caseydavenport** [comment]
@Mutantt have you tried doing a TCP dump of the pod interface to see what packets are being sent? That would help diagnose the issue a bit further. It would be useful to see what the pod is trying to do, and what (if any) response it is getting from the node.

**Comment by Mutantt** [comment]
@caseydavenport Here is the information that you requested: 

root@# tcpdump -i calif8322348b74 -vvv -e -nn 
tcpdump: listening on calif8322348b74, link-type EN10MB (Ethernet), capture size 262144 bytes
11:17:01.071599 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35054, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 1, length 64
11:17:02.092758 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35126, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 2, length 64
11:17:03.116728 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35366, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 3, length 64
11:17:04.140735 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35404, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 4, length 64
11:17:05.164749 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35486, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 5, length 64
11:17:06.188690 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype ARP (0x0806), length 42: Ethernet (len 6), IPv4 (len 4), Request who-has 169.254.1.1 tell 10.233.110.227, length 28
11:17:06.188751 ee:ee:ee:ee:ee:ee > 82:14:1f:8e:44:1d, ethertype ARP (0x0806), length 42: Ethernet (len 6), IPv4 (len 4), Reply 169.254.1.1 is-at ee:ee:ee:ee:ee:ee, length 28
11:17:06.188783 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35729, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 6, length 64
11:17:07.212766 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 35832, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 7, length 64
11:17:08.236723 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 36036, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 8, length 64
11:17:09.260705 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 36044, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 9, length 64
11:17:10.284762 82:14:1f:8e:44:1d > ee:ee:ee:ee:ee:ee, ethertype IPv4 (0x0800), length 98: (tos 0x0, ttl 64, id 36249, offset 0, flags [DF], proto ICMP (1), length 84)
    10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 10, length 64

**Comment by caseydavenport** [comment]
ICMP requests are being sent which means the ARP went through successfully. 

I wonder if the packets are being dropped due to a firewall rule on the host?

**Comment by Mutantt** [comment]
@caseydavenport I sent tcpdump of node. I pinged from pod and tcpdump on corresponding node. Is it what you want?

**Comment by caseydavenport** [comment]
Yep, that's what I asked for. However, there's nothing obviously wrong with ARP in that capture. 

> 10.233.110.227 > 172.29.2.22: ICMP echo request, id 91, seq 9, length 64

Just to double-check, who is 172.29.2.22 in this scenario? The node hosting that pod?

I think the next step is to identify what is happening to the packets next. Could you check `iptables-save -c | grep DROP` on the host and see if any of the rules are being hit?

**Comment by Mutantt** [comment]
@caseydavenport Shouldn't there be two pairs of ARP request & reply?one pair for pod and one pair for node?when I ping the node from pod and check ARP cache with "ip neighbour",there is a record with my pod ip and the state is STALE,but when I ping the pod from my node the state change to REACHABLE and only then ping from pod is successful.

**Comment by Mutantt** [comment]
Hi @caseydavenport 
Sorry for the late reply
when I ping any pod that has been scheduled to specific worker from that specific worker and doing tcpdump on the pod I see two pairs of ARP request & reply:
----
listening on any, link-type LINUX_SLL2 (Linux cooked v2), snapshot length 262144 bytes
19:59:13.649719 eth0  B   ARP, Request who-has net-tools-6476cf44bd-4xxk8 tell 172-29-2-23.csi-rbdplugin-metrics.rook-ceph.svc.cluster.local, length 28
19:59:13.649733 eth0  Out ARP, Reply net-tools-6476cf44bd-4xxk8 is-at ca:4b:13:f8:b9:33 (oui Unknown), length 28
19:59:13.649738 eth0  In  IP 172-29-2-23.csi-rbdplugin-metrics.rook-ceph.svc.cluster.local > net-tools-6476cf44bd-4xxk8: ICMP echo request, id 10, seq 1, length 64
19:59:13.649753 eth0  Out ARP, Request who-has 169.254.1.1 tell net-tools-6476cf44bd-4xxk8, length 28
19:59:13.649755 eth0  In  ARP, Reply 169.254.1.1 is-at ee:ee:ee:ee:ee:ee (oui Unknown), length 28
----
but I see one pair of ARP request and reply when I ping the worker from that pod.I sent it's tcpdump before.
please advise if I've missed anything

**Comment by caseydavenport** [comment]
The first request / reply appears to be from the host to the pod, asking for its mac address, which is then returned.
The second request / reply appears to be the reverse - the pod doing a lookup of its default gateway. 

If you're only seeing the pod lookup its default gateway in the pod->host scenario, this could be because the host is populating its arp cache from the received packet, or it could potentially mean that the host is failing to ARP when trying to send the reply packet.

One thing you might check would be to flush the host's ARP table, snapshot it, and then ping from the pod to the host, snapshot the arp table again, and compare.

e.g., 

- flush: `ip -s -s neigh flush all`
- snapshot: `ip neigh > arp-before.txt`
- ping
- snapshot `ip neigh > arp-after.txt`

Might tell us if the cache is being populated without sending an ARP packet.

**Comment by Mutantt** [comment]
@caseydavenport thank you for you reply
these are the result:

[arp-before.txt](https://github.com/projectcalico/calico/files/7639172/arp-before.txt)
[arp-after.txt](https://github.com/projectcalico/calico/files/7639171/arp-after.txt)

The pod's IP is 10.233.88.208 and the host's IP is 172.29.2.23
After I pinged the host from the pod,the record appeared but was in STALE state.It remains in that state until I ping the pod from the host.

**Comment by Mutantt** [comment]
After several weeks of struggling with this issue,finally we found out what cause this. we used Kubespray to deploy our k8s cluster. It has a parameter called  calico_endpoint_to_host_action.This variable takes accept and return values. The default value is "accept" but I don't know why it was set to return in our cluster. We changed it to accept and everything is ok for now.
@caseydavenport  thank you for your help.
